### PR TITLE
Simple Clean up, .gitignore improved, PR template added

### PR DESCRIPTION
**Pull request by smttsp**
## Summary

> What is the reason for this Pull-Request? 

This PR improves the .gitignore as well as adds a PR template to be filled in any PR created. 

Added the following sections: 
- Byte-compiled / optimized / DLL files
- development environment configuration files (`.idea, .vscode`)
- Python Virtual Environments (common ones: venv, env, .direnv etc)
- C/c++ extensions (.so)
- Distribution / packaging (.egg, lib, lib64, wheels etc)
- Installer logs
- Unit test / coverage reports
- log files
- IPython (e.g. profile_default/)
- mkdocs documentation
- mypy caches
- compressed files (zip, gz, tar, 7z etc)
- other files (output, wandb)

## Impact

> What is the scope of your changes and who/where do you think they will affect?

- improved gitignore
- new template

## Testing

> What testing have you done and verified? If you can't test, please say why that is.

- I have tested both locally
- I am using the template now

## Ticket(s) or Conversations

> What Git or Slack items (Issues, threads etc) are specifically related to this work? Please link them here.

-N/A

## References

> Do you have any relevant URLs or Wiki pages that could help give background information for the changes in this Pull-Request? These are usually StackOverflow posts, Notion wiki pages, Wikipedia articles, or links to APIs and Documentation of third-party packages.

-

### CODE REVIEW

**Review by smttsp — COMMENTED** [review]
ready for review

File: .github/pull_request_template.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
+## Summary
+
+> What is the reason for this Pull-Request? Are you bumping the version?
```
Comment (jwmueller) [inline .github/pull_request_template.md:3]:
```suggestion
> What is the reason for this Pull-Request?
```

File: .github/pull_request_template.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+## Impact
+
+> What is the scope of your changes and who/where do you think they will affect?
+
+
+
+## Testing
+
+> What testing have you done and verified? If you can't test, please say why that is.
```
Comment (jwmueller) [inline .github/pull_request_template.md:15]:
```suggestion
> What testing have you done and verified? Please link to the one unit test that is most end-to-end check of your new functionality. It's ok if your unit tests are not yet comprehensive when you first open the PR, we can revisit them later!  
```

File: .github/pull_request_template.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+-
+
+## Ticket(s) or Conversations
+
+> What Git or Slack items (Issues, threads etc) are specifically related to this work? Please link them here.
+
+
+## References
+
+> Do you have any relevant URLs or Wiki pages that could help give background information for the changes in this Pull-Request? These are usually StackOverflow posts, Notion wiki pages, Wikipedia articles, or links to APIs and Documentation of third-party packages.
```
Comment (jwmueller) [inline .github/pull_request_template.md:26]:
```suggestion
## References

> Include any extra information that may be helpful to reviewers of your Pull-Request here (e.g. relevant URLs or Wiki pages that could help give background information). If relevant, please include additional code snippets (and their outputs/return values) showing alternative ways to use your newly added methods.
```

File: .gitignore [hunk 4]
```diff
@@ -21,3 +20,135 @@ image_files*
...
+*$py.class
+.direnv
+.idea
+.idea/**
+.vscode
+.vscode/**
+
+# C extensions
+*.so
+**.txt
```
Comment (jwmueller) [inline .gitignore:35]: why are you proposing to add: **.txt ?
That seems odd to ignore to me

File: .gitignore [hunk 5]
```diff
@@ -1,4 +1,3 @@
-.idea/
 .DS_Store
 __pycache__/
```
Comment (jwmueller) [inline .gitignore:2]: why did you remove `*/.DS_Store` that used to be here?

More generally, why did you remove certain things and add certain things in .gitignore. The underlying reasoning is non-obvious to me as an observer

Comment (smttsp) [reply .gitignore:2]: I moved it down as there was a duplicate, but I can move it up if you want. I thought it would be better to be grouped together with relevant things

Several of them really should be added, for example `ENV` without which I get about 30K files suggested to be committed. Similarly, I removed `direnv`, `venv` and other common variations envs as they are typically causing complications.

File: .gitignore (hunk 4 above)
Comment (smttsp) [reply .gitignore:35]: I copied them from my other repos, but I can remove some of them if you want. I can delete a bunch of other things that are more related to personal choices.

File: .gitignore (hunk 5 above)
Comment (smttsp) [reply .gitignore:2]: I don't think I deleted any of those tbh. I moved `.idea` down.  DS_store is also still there. (without *). 

>data/
tests/fasttext_data/
build (not build/)
cleanlab-docs
/docs/source/notebooks/spoken_digits

Comment (smttsp) [reply .gitignore:2]: > data/
tests/fasttext_data/
build (not build/)
cleanlab-docs
/docs/source/notebooks/spoken_digits

looking at the git history, I don't think they were ever added to this repo

File: .gitignore (hunk 4 above)
Comment (jwmueller) [reply .gitignore]: making some changes based on personal preferences is fine! But it will be helpful to clarify which additions & deletions are based on personal preferences VS which additions & deletions are based on improving the code organization

File: .gitignore (hunk 5 above)
Comment (jwmueller) [reply .gitignore:2]: Doh I think I was comparing against the wrong .gitignore file, please ignore the specific list here. 

The general principle is just to clarify what things are being added/deleted in your PR, and when these are based on personal preferences VS based on better code organization.

File: .github/pull_request_template.md (hunk 1 above)
Comment (jwmueller) [reply .github/pull_request_template.md]: I don't think this suggestion got incorporated in the latest version? 
This is most important part of the PR template IMO, it saves immense time for reviewers to see code snippet and outputs at the top of the PR before digging into details.

**Comment by jwmueller** [comment]
@smttsp once you've finalized .github/pull_request_template.md

it would be awesome if you can also make a PR with the same template to cleanlab/cleanlab repo!

**Review by jwmueller — APPROVED** [review]
There was one overlooked suggestion, which you can commit, and then this LGTM. 

Leaving final review + merge to @sanjanag who has relied on the .gitignore more than any other developer of this repo thus far.

File: .github/pull_request_template.md (hunk 1 above)
Comment (smttsp) [reply .github/pull_request_template.md]: Interesting! 

I am pretty sure I committed the suggestion but somehow it is gone. 

I am recommitting it now.

Comment (smttsp) [reply .github/pull_request_template.md]: ✅ Updated the PR template with the above suggestion.

**Comment by smttsp** [comment]
@jwmueller, would you like me to update gitignore as well?

**Review by sanjanag — COMMENTED** [review]
Overall, please remove all unnecessary file escapes from the gitignore, it's hard to comment at each one of them.

**Comment by smttsp** [comment]
@sanjanag if there is anything in gitignore you wanna keep, I can keep them. Otherwise I am dropping all the gitignore related commits.

**Comment by sanjanag** [comment]
Hi @smttsp ! Your PR template looks good, We don't want to clutter the gitignore with supposedly "common stuff" that does not concern this repo. I'd recommend you only keep the following sections in the gitignore and remove the rest. Thank you.
```
# development environment configuration files
# Python Environments
# Distribution / packaging
# Data files
```

**Comment by smttsp** [comment]
@sanjanag would you like me to have the same exact changes as this one: https://github.com/cleanlab/cleanlab/pull/867/files?

**Comment by sanjanag** [comment]
Yes that looks good @smttsp

**Comment by smttsp** [comment]
> Yes that looks good @smttsp. Please make sure to not remove any existing lines, unless they are duplicated.

@sanjanag, I have copied it. Now, all the .gitignore lines in cleanlab are here as well, I haven't deleted anything from here either.

File: .gitignore [hunk 6]
```diff
@@ -1,23 +1,158 @@
...
+lib64/
+parts/
+sdist/
+var/
+wheels/
 *.egg-info/
-build/
+.installed.cfg
+*.egg
+MANIFEST
```
Comment (sanjanag) [inline .gitignore:27]: Does this line exclude MANIFEST.in?

File: .gitignore [hunk 7]
```diff
@@ -1,23 +1,158 @@
...
+.tox/
 .coverage
+.coverage.*
+.cache
+nosetests.xml
 coverage.xml
+*.cover
+.hypothesis/
+.pytest_cache/
+.coverage*
```
Comment (sanjanag) [inline .gitignore:40]: Does this line exclude .coveragerc?

File: .gitignore [hunk 8]
```diff
@@ -1,23 +1,158 @@
...
+.pytest_cache/**
 
 # Misc
 results/
 image_files*
 
 # datasets
 cifar*
+
+cleanlab/*.ipynb
```
Comment (sanjanag) [inline .gitignore:52]: this is not valid in this repo, please remove this

File: .gitignore [hunk 9]
```diff
@@ -1,23 +1,158 @@
...
+tests/fasttext_data/
```
Comment (sanjanag) [inline .gitignore:53]: this is not valid in this repo, please remove this

File: .gitignore [hunk 10]
```diff
@@ -1,23 +1,158 @@
...
+data/
+docs/_build/
+
+# Byte-compiled / optimized / DLL files
+__pycache__/
+*.py[cod]
+*$py.class
+
+# C extensions
+*.so
```
Comment (sanjanag) [inline .gitignore:63]: this is not valid in this repo, please remove this

File: .gitignore [hunk 11]
```diff
@@ -1,23 +1,158 @@
...
+
+
+# PyInstaller
```
Comment (sanjanag) [inline .gitignore:66]: this is not valid in this repo, please remove this

File: .gitignore [hunk 12]
```diff
@@ -1,23 +1,158 @@
...
+# Rope project settings
+.ropeproject
+
+# mkdocs documentation
+/site
+
+# mypy
+.mypy_cache/
+
+# Downloaded datasets for docs
```
Comment (sanjanag) [inline .gitignore:136]: This section is not valid in this repo, please remove this

File: .gitignore [hunk 13]
```diff
@@ -1,23 +1,158 @@
...
+ENV/
+env.bak/
+venv.bak/
+.direnv
+
+# Spyder project settings
+.spyderproject
+.spyproject
+
+# Rope project settings
```
Comment (sanjanag) [inline .gitignore:127]: This is not valid in this repo, please remove this

File: .gitignore [hunk 14]
```diff
@@ -1,23 +1,158 @@
...
+.env
+.venv
+env/
+venv/
+ENV/
+env.bak/
+venv.bak/
+.direnv
+
+# Spyder project settings
```
Comment (sanjanag) [inline .gitignore:123]: This is not valid in this repo, please remove this

File: .gitignore [hunk 15]
```diff
@@ -1,23 +1,158 @@
...
+target/
+
+# Jupyter Notebook
+.ipynb_checkpoints
+
+# pyenv
+.python-version
+
+# celery beat schedule file
+celerybeat-schedule
```
Comment (sanjanag) [inline .gitignore:108]: This is not valid in this repo, please remove this

File: .gitignore [hunk 16]
```diff
@@ -1,23 +1,158 @@
...
+
+# SageMath parsed files
```
Comment (sanjanag) [inline .gitignore:110]: This is not valid in this repo, please remove this

File: .gitignore [hunk 17]
```diff
@@ -1,23 +1,158 @@
...
+
+# Spyder project settings
+.spyderproject
+.spyproject
+
+# Rope project settings
+.ropeproject
+
+# mkdocs documentation
+/site
```
Comment (sanjanag) [inline .gitignore:131]: This is not valid in this repo, please remove this

File: .gitignore [hunk 18]
```diff
@@ -1,23 +1,158 @@
...
+# Downloaded datasets for docs
+/docs/source/notebooks/aclImdb
+/docs/source/notebooks/data
+/docs/source/notebooks/*.gz
+/docs/source/notebooks/spoken_digits
+/docs/source/notebooks/pretrained_models
+/docs/source/tutorials/datalab/datalab-files/
+
+# IPython
+profile_default/
```
Comment (sanjanag) [inline .gitignore:145]: This is not valid in this repo, please remove this

File: .gitignore [hunk 19]
```diff
@@ -1,23 +1,158 @@
...
+cleanlab-docs
+
+# PyBuilder
+target/
+
+# Jupyter Notebook
+.ipynb_checkpoints
+
+# pyenv
+.python-version
```
Comment (sanjanag) [inline .gitignore:105]: This is not valid in this repo, please remove this

File: .gitignore [hunk 20]
```diff
@@ -1,23 +1,158 @@
...
+# Scrapy stuff:
+.scrapy
+
+# Sphinx documentation
+build
+_build
+cleanlab-docs
+
+# PyBuilder
+target/
```
Comment (sanjanag) [inline .gitignore:99]: This is not valid in this repo, please remove this

File: .gitignore [hunk 21]
```diff
@@ -1,23 +1,158 @@
...
+instance/
+.webassets-cache
+
+# Scrapy stuff:
+.scrapy
+
+# Sphinx documentation
+build
+_build
+cleanlab-docs
```
Comment (sanjanag) [inline .gitignore:96]: This is not valid in this repo, please remove this

File: .gitignore [hunk 22]
```diff
@@ -1,23 +1,158 @@
...
+*.mo
+*.pot
+
+# Django stuff:
+*.log
+local_settings.py
+db.sqlite3
+logs/
+
+# Flask stuff:
```
Comment (sanjanag) [inline .gitignore:86]: This is not valid in this repo, please remove this

File: .gitignore [hunk 23]
```diff
@@ -1,23 +1,158 @@
...
+
+# Installer logs
+pip-log.txt
+pip-delete-this-directory.txt
+
+# Translations
+*.mo
+*.pot
+
+# Django stuff:
```
Comment (sanjanag) [inline .gitignore:80]: This is not valid in this repo, please remove this

File: .gitignore [hunk 24]
```diff
@@ -1,23 +1,158 @@
...
+#  Usually these files are written by a python script from a template
+#  before PyInstaller builds the exe, so as to inject date/other infos into it.
+*.manifest
+*.spec
+
+# Installer logs
+pip-log.txt
+pip-delete-this-directory.txt
+
+# Translations
```
Comment (sanjanag) [inline .gitignore:76]: This is not valid in this repo, please remove this

File: .gitignore [hunk 25]
```diff
@@ -1,23 +1,158 @@
...
+*.so
+
+
+# PyInstaller
+#  Usually these files are written by a python script from a template
+#  before PyInstaller builds the exe, so as to inject date/other infos into it.
+*.manifest
+*.spec
+
+# Installer logs
```
Comment (sanjanag) [inline .gitignore:72]: This is not valid in this repo, please remove this

**Comment by sanjanag** [comment]
Hi @smttsp ! I have added a few more comments, could you please address them? Thanks.

File: .gitignore (hunk 6 above)
Comment (smttsp) [reply .gitignore:27]: No, it doesnt exclude

File: .gitignore (hunk 7 above)
Comment (smttsp) [reply .gitignore:40]: yes, it excludes

File: .gitignore (hunk 8 above)
Comment (smttsp) [reply .gitignore:52]: sure

File: .gitignore (hunk 9 above)
Comment (smttsp) [reply .gitignore:53]: sure

File: .gitignore (hunk 10 above)
Comment (smttsp) [reply .gitignore:63]: many of these are very common stuff in gitignore, in fact, most of these lines are coming from Github's default Python gitignore. 

I found being proactive much more effective than being retroactive. We will add common things now, but we won't need to add them to gitignore if/when it happens.

Note that prior to this PR, when I set up my environment, I was seeing 30K files to be committed as the `ENV` (default) directory wasn't in gitignore.

File: .gitignore (hunk 8 above)
Comment (smttsp) [inline .gitignore:52]:
```suggestion
```

File: .gitignore (hunk 9 above)
Comment (smttsp) [inline .gitignore:53]:
```suggestion
```

File: .gitignore (hunk 21 above)
Comment (smttsp) [inline .gitignore:96]:
```suggestion
```

File: .gitignore (hunk 19 above)
Comment (smttsp) [inline .gitignore:105]:
```suggestion
```

File: .gitignore (hunk 15 above)
Comment (smttsp) [inline .gitignore:108]:
```suggestion
```

File: .gitignore [hunk 26]
```diff
@@ -1,23 +1,158 @@
...
+# mypy
+.mypy_cache/
+
+# Downloaded datasets for docs
+/docs/source/notebooks/aclImdb
+/docs/source/notebooks/data
+/docs/source/notebooks/*.gz
+/docs/source/notebooks/spoken_digits
+/docs/source/notebooks/pretrained_models
+/docs/source/tutorials/datalab/datalab-files/
```
Comment (smttsp) [inline .gitignore:142]:
```suggestion
```

File: .gitignore (hunk 6 above)
Comment (sanjanag) [reply .gitignore:27]: can you still remove it? I think it's confusing since we have a MANIFEST.in file

File: .gitignore (hunk 7 above)
Comment (sanjanag) [reply .gitignore:40]: We don't want to exclude it, can you please remove it?

**Review by sanjanag — APPROVED** [review]
